api/common/modelstatus.go

```diff
@@ -37,8 +45,40 @@ func (c *ModelStatusAPI) ModelStatus(tags ...names.ModelTag) ([]base.ModelStatus
 		return nil, err
 	}
 
-	results := make([]base.ModelStatus, len(result.Results))
-	for i, r := range result.Results {
+	if olderVersion {
+		// This caters for the fact that older version of base.ModelStatus
+		// did not have an Error field.
+		return c.processModelStatusResultsV3(result.Results)
+	}
+
+	return c.processModelStatusResults(result.Results)
+}
+
+func (c *ModelStatusAPI) processModelStatusResults(rs []params.ModelStatus) ([]base.ModelStatus, error) {
+	results := make([]base.ModelStatus, len(rs))
+	for i, r := range rs {
+		if r.Error != nil {
+			results[i].Error = errors.Trace(r.Error)
+			continue
+		}
+		model, err := names.ParseModelTag(r.ModelTag)
+		if err != nil {
+			results[i].Error = errors.Annotatef(err, "model tag %v", r.ModelTag)
+			continue
+		}
+		owner, err := names.ParseUserTag(r.OwnerTag)
+		if err != nil {
+			results[i].Error = errors.Annotatef(err, "owner tag %v", r.OwnerTag)
+			continue
+		}
+		results[i] = constructModelStatus(model, owner, r)
+	}
+	return results, nil
+}
+
+func (c *ModelStatusAPI) processModelStatusResultsV3(rs []params.ModelStatus) ([]base.ModelStatus, error) {
```

Comment:
I don't understand why we need different behaviour here for V3 on the api (not apiserver) side.
The api layer can call either v3 or v4 and still present the new interface to callers. The []base.ModelStatus result slice can still be filled in with an error for a result if needed.
eg,
even for v3 backend, when we iterate over the rs slice below, if we get a model tag parse error we can fill in the result error and loop to the next record

This works because this api layer facade is used by all the code we control so any call sites can be updated to the new api layer semantics.

Reply:
removed. Thank you for the explanation ;)